File: Frontend/app/src/app/base/services/base.service.ts
import { Page } from '../models/page';
import { HttpClient } from '@angular/common/http';
import { DataPage } from '../models';

export class BaseService<TModel> {
  resource!: string;
  url!: string;

  constructor(protected http: HttpClient) { }

  getData(page: Page = {} as any, query: string = '') {
    const queryString = query ? `?query=${query}` : '';
    return this.http
      .post(`${this.url}/GetData${queryString}`, page)
      .toPromise() as Promise<DataPage<TModel>>;
  }

  deleteMultiple(listId: Array<string>) {
    const data = { ids: listId }
    return this.http
      .post(`${this.url}/DeleteMultiple`, data)
      .toPromise() as Promise<boolean>;
  }

  delete(id: string) {
    return this.http
      .delete(`${this.url}/${id}`)
      .toPromise() as Promise<boolean>;
  }

  getById(id: string) {
    return this.http
      .get(`${this.url}/${id}`)
      .toPromise() as Promise<TModel>;
  }

  insert(data: TModel) {
    return this.http
      .post(`${this.url}`, data)
      .toPromise() as Promise<TModel>;
  }

  update(id: string, data: TModel) {
    return this.http
      .put(`${this.url}/${id}`, data)
      .toPromise() as Promise<TModel>;
  }
}

File: Backend/API/API/Services/BaseService.cs
using API.DTOs;
using API.Exceptions;
using API.Extensions;
using API.Model;
using API.Repository.Interface;
using API.Services.Interface;

namespace API.Services
{
    public class BaseService<TRepository, TModel> : IBaseService<TModel>
        where TRepository : IBaseRepository<TModel>
        where TModel : BaseModel
    {
        protected readonly IBaseRepository<TModel> _repository;

        public BaseService(IBaseRepository<TModel> repository)
        {
            _repository = repository;
        }

        public virtual async Task<bool> Delete(Guid id)
        {
            var exist = await _repository.Exists(id);
            if (!exist)
            {
                throw new HttpResponseException(StatusCodes.Status400BadRequest, "ID is invalid");
            }

            return await _repository.Delete(id);
        }

        public virtual async Task<DataPage> GetData(Page page, string query = null)
        {
            var dataQuery = _repository.Get();
            if (!string.IsNullOrEmpty(query))
                dataQuery = dataQuery.Where(r => r.Name.Contains(query));

            var data = await dataQuery.GetPage(page);
            var result = new DataPage
            {
                Data = data,
                Page = page
            };
            return result;
        }

        public virtual async Task<TModel> GetById(Guid id)
        {
            var result = await _repository.GetById(id);
            if (result is null)
            {
                throw new HttpResponseException(StatusCodes.Status400BadRequest, "ID is invalid");
            }

            return result;
        }

        public virtual async Task<TModel> Insert(TModel model)
        {
            return await _repository.Insert(model);
        }

        public virtual async Task<TModel> Update(Guid id, TModel model)
        {
            var exist = await _repository.Exists(id);
            if (!exist)
            {
                throw new HttpResponseException(StatusCodes.Status400BadRequest, "ID is invalid");
            }

            return await _repository.Update(model);
        }

        public virtual async Task<bool> DeleteMultiple(ListID listID)
        {
            return await _repository.DeleteMultiple(listID.Ids);
        }

        public virtual async Task<List<TModel>> GetMultiple(ListID listID)
        {
            return await _repository.GetMultiple(listID.Ids);
        }
    }
}